### ISSUE TITLE
Unnecessary return added setting `always_use_return = true`.

### ISSUE THREAD

**Issue by kellertuer** [domluna/JuliaFormatter.jl#233]
I think when formatting the following function
```julia
function f(a)
    if a > 0
        return -1
    else
        return 1
    end
end
```
using the option `always_use_return = true` we get
```julia
function f(a)
    return if a > 0
        return -1
    else
        return 1
    end
end
```
I was surprised this function still works as intended, but I think the return might be too much in the second line. This might be hard to track down, but is here the always maybe treated a little to strict? Basically the last two lines in the cases do have return values, so the first form should be fine?

**Comment by domluna** [comment]
Right now only the top-level is checked to contain `return`. I wonder if it would be sufficient to merely check if `return` is present at any point in the underlying tree. For practical use cases I believe this would be true, but there could be weird corner cases.

**Comment by kellertuer** [comment]
The corner cases might indeed be challenging. What I am not sure about is, _if_ the second branch, say, does not have a return-statement, does the addd turn value change the behaviour?

Despite that `return if` just does not look right (nor stylish).

**Comment by StefanKarpinski** [comment]
In general, I think you want to insert the return only if there's some way for control flow to come back. So, if all branches of a conditional have return or throw (or heuristically, `error`) on them, then no need to insert a `return`. Likewise, if you have a loop and it always returns, then you don't need the outer return—but that's very hard to know. You also can't generally know if a function call returns or not, but containing a literal `throw` or `error` is probably a good enough heuristic.

**Comment by StefanKarpinski** [comment]
In terms of the always-returning loop case, suppose we have this loop where we're sure the return occurs at some point:
```jl
function search(needle, haystack)
    for (i, value) in enumerate(haystack)
        value == needle && return i
    end
end
```
Even if I'm sure that this is supposed to return, I would defensively write the code like this:
```jl
function search(needle, haystack)
    for (i, value) in enumerate(haystack)
        value == needle && return i
    end
    @assert false "index not found"
end
```
Or maybe with a normal error; The reason to use `@assert false "message"` is to indicate that it's an internal logical error if that point in code is reached. In any case, since loops evaluate to nothing, the original version could be formatted to:
```jl
function search(needle, haystack)
    for (i, value) in enumerate(haystack)
        value == needle && return i
    end
    return
end
```
You could know not to insert a `return` in the `@assert false` version by knowing that it's an unconditional error exit (special case `throw`, `error` and `@assert false` I guess).

Bottom line, I guess the rule would be to insert an explicit `return` unless every basic block has `return`, `throw`, `error` or `@assert false` on it. For a loop in return position, you just put `return` after the loop since the loop evaluates to `nothing`; for a conditional in return position, you put the `return` in front of the entire conditional expression. Do you have the machinery to do basic block analysis on code? That's kind of tricky but it's already done by the compiler.

**Comment by eschnett** [comment]
For the record (it might be implied by the comments above):
- When the last statement of a function is a `for` loop, it is changed to `return for`. For loops don't return a value, hence there should be a plain `return` after the `for` loop instead.
- `@assert` is changed to `return @assert`.

**Comment by domluna** [comment]
I've restricted the cases which this rule is applicable to #275. This seems to be easiest workaround for now.
